### PR TITLE
fix(ngOptions): remove selected attribute from unselected options

### LINKED ISSUE #14419: "Selected" attributes for <select> options elements not being updated properly
**Request Type:** bug

**Current Behavior:**
When a select element is created with both ng-options and ng-model attributes to track the selected option, the "selected" element property is added to the selected attribute (per this issue: #8366). However, when a different option is selected, the "selected" attribute is not removed, resulting in all of the options containing a selected="selected" attribute (if all options have been selected). This is important due to accessibility guidelines stipulate the use of the selected attribute on options to indicate what option has been selected. [Aria Guidelines] (http://www.w3.org/TR/WCAG20-TECHS/H91).

**Reproduce?**
Any select box that has a ng-model tracking the value within an ng-options attribute for a select element. This is a modification of the Angular docs example for ng-options that illustrates nicely the select attribute not being added and removed properly: https://plnkr.co/edit/PjbWbo

**Expected behavior?**
Only one selected="selected" attributes should be present within the options of a select element. When a selection is changed the attribute should be removed and added to the newly selected option.

**What is the motivation / use case for changing the behavior?**
As per mentioned above, having one option with a selected attribute allows for proper accessibility functionality.

**Versions**
This was tested in Angular 1.4.1 and 1.5.3 in Chrome 49, Canary 52, Safari 9 and Firefox 42

I found this was addressed in commits 79538af and 3d393dc

### LINKED ISSUE #14892: Select ng-options is dont reset with '<input type="reset"> '
***Note*: for support questions, please use one of these channels: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#question. This repository's issues are reserved for feature requests and bug reports.**

**Do you want to request a *feature* or report a *bug*?**
Bug

**What is the current behavior?**
When I have a select with ng-options with an array with objects, the behavior is weird when I reset the form by clicking in a  `<input type="reset" value="clear">`

The problem occurs because when you select an option angular put a attribute `selected="selected"` on the option and never clears the attribute in the others options.

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem via https://plnkr.co or similar (template: http://plnkr.co/edit/tpl:yBpEi4).**

Chosse different options in the select and then press the reset button.

You can watch the error here: https://plnkr.co/edit/Epq5fVrS0I6p2csrntli?p=preview

select-dom

**What is the expected behavior?**
The standard reset form button works properly

**What is the motivation / use case for changing the behavior?**
The standard reset form button doesn't works properly when ng-options is used with a array with objects.

**Which versions of Angular, and which browser / OS are affected by this issue? Did this work in previous versions of Angular? Please also test with the latest stable and snapshot (https://code.angularjs.org/snapshot/) versions.**
1.5.5, 1.5.7

**Other information (e.g. stacktraces, related issues, suggestions how to fix)**
The partial solution of this issue is commenting the line of the angular.js 1.5.5

28012  `// option.element.setAttribute('selected', 'selected')`;

### PR DESCRIPTION
**Pull request by Narretz**
**What kind of change does this PR introduce? (Bug fix, feature, docs update, ...)**
bug fix

**What is the current behavior? (You can also link to an open issue here)**
See #14892 nad #14419

**Does this PR introduce a breaking change?**
No

**Please check if the PR fulfills these requirements**
- [x] The commit message follows our guidelines: https://github.com/angular/angular.js/blob/master/CONTRIBUTING.md#commit-message-format
- [x] Tests for the changes have been added (for bug fixes / features)
- ~~[ ] Docs have been added / updated (for bug fixes / features)~~

**Other information**:

When the select model changes, we add the "selected" attribute to the selected option, so that
screen readers know which option is selected.

Previously, we failed to remove the attribute when the model changed to match a different option, or
the unknown / empty option.

When using "track by", the behavior would also show when a user selected an option, and then the
model was changed, because track by watches the tracked expression, and calls the $render function
on change.

This fix reads the current select value, finds the matching option and removes the "selected"
attribute.

Fixes #14892
Fixes #14419
Related #12731

### CODE REVIEW

File: test/ng/directive/ngOptionsSpec.js [hunk 1]
```diff
@@ -743,6 +743,37 @@ describe('ngOptions', function() {
...
 
+  it('should remove the "selected" attribute from the previous option when the model changes', function() {
+    scope.values = [{id: 10, label: 'ten'}, {id:20, label: 'twenty'}];
+
+    createSelect({
+      'ng-model': 'selected',
+      'ng-options': 'item.label for item in values'
+    }, true);
+
+    var resetButton = angular.element('<input type="reset" />');
```
Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: Isn't the reset button irrelevant. If so, I would avoid complicating the testcase.

File: test/ng/directive/ngOptionsSpec.js [hunk 2]
```diff
@@ -743,6 +743,37 @@ describe('ngOptions', function() {
...
+    formElement.append(resetButton);
+
+    scope.selected = scope.values[0];
+    scope.$digest();
+
+    var options = element.find('option');
+    expect(options[1].getAttribute('selected')).toBe('selected');
```
Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: I would feel more comfortable, if you checked all options for the `selected` attribute (here and below).

File: test/ng/directive/ngOptionsSpec.js [hunk 3]
```diff
@@ -1395,6 +1426,28 @@ describe('ngOptions', function() {
         });
       }).not.toThrow();
     });
+
+    it('should remove the "selected" attribute when the model changes', function() {
```
Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: _when the model changes_  -->  _when the view changes_ (?)

File: test/ng/directive/ngOptionsSpec.js [hunk 4]
```diff
@@ -1395,6 +1426,28 @@ describe('ngOptions', function() {
...
+
+      var options = element.find('option');
+      browserTrigger(options[2], 'click');
+
+      expect(scope.selected).toEqual(scope.arr[1]);
+
+      scope.selected = {};
+      scope.$digest();
+
+      options[2].selected = false;
```
Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: Why is this necessary?

Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: Sorry, left over from a test inside the test

File: test/ng/directive/ngOptionsSpec.js (hunk 3 above)
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: Technically, the problem appears when you update the model (which then updates the view)

File: test/ng/directive/ngOptionsSpec.js [hunk 5]
```diff
@@ -743,6 +743,37 @@ describe('ngOptions', function() {
     expect(options.eq(3)).toEqualOption('c');
   });
 
+  it('should remove the "selected" attribute from the previous option when the model changes', function() {
+    scope.values = [{id: 10, label: 'ten'}, {id:20, label: 'twenty'}];
+
+    createSelect({
+      'ng-model': 'selected',
+      'ng-options': 'item.label for item in values'
```
Comment (fovotz) [inline test/ng/directive/ngOptionsSpec.js]: I think is missing the "track by id", sorry if I wrong

Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: There's a different test for track by

File: test/ng/directive/ngOptionsSpec.js (hunk 2 above)
Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: Good catch, because "selected" isn't removed from the empty option

Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: It's based on personal experience. I tried a fix for this myself and I missed the empty option the first time :smile:

Comment (Narretz) [inline test/ng/directive/ngOptionsSpec.js]: Ah, sorry I should have assigned myself at the beginning, so we didn't have to duplicate work

Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: I can't blame, since I did the same :stuck_out_tongue:

**Comment by Narretz** [comment]
Sorry for the potential ton of update emails you got, @gkalpak But now it's ready for another review

File: test/ng/directive/ngOptionsSpec.js [hunk 6]
```diff
@@ -744,6 +778,42 @@ describe('ngOptions', function() {
...
+    scope.$digest();
+
+    expect(options[0]).not.toBeMarkedAsSelected();
+    expect(options[1]).not.toBeMarkedAsSelected();
+    expect(options[2]).toBeMarkedAsSelected();
+
+    scope.selected = 'no match';
+    scope.$digest();
+
+    expect(options[0].selected).toBe(true);
```
Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: Is this necessary? Isn't `.toBeMarkedAsSelected()` enough?

File: src/ng/directive/ngOptions.js [hunk 7]
```diff
@@ -467,9 +467,15 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
...
         selectCtrl.writeValue = function writeNgOptionsValue(value) {
+          var selectedOption = options.selectValueMap[selectElement.val()];
           var option = options.getOptionFromViewValue(value);
 
+          // Make sure to remove the selected attribute from the previously selected option
+          // Otherwise, screen readers might get confused
+          if (selectedOption) selectedOption.element.removeAttribute('selected');
+
           if (option) {
+            if (providedEmptyOption) emptyOption.removeAttr('selected');
```
Comment (gkalpak) [inline src/ng/directive/ngOptions.js]: I think ti would be safer to move this inside the `removeEmptyOption()` function, s we don't have to manually call it every time.

For example, there is currently a call for `removeEmptyOption()` inside `selectCtrl.readValue()`. Do we need to also remove the `selected` attribute? Probably not, but who knows :stuck_out_tongue:

**Comment by gkalpak** [comment]
A couple of minor comments/questions. Otherwise LGTM :+1:

File: src/ng/directive/ngOptions.js (hunk 7 above)
Comment (Narretz) [inline src/ng/directive/ngOptions.js]: I guess not, because I think when the empty option is removed and then re-added, we need it to be selected. But moving it into the fn it saves precious CPU cycle!

Comment (Narretz) [inline src/ng/directive/ngOptions.js]: I removed this line completely and all the tests pass ... not good

File: src/ng/directive/ngOptions.js [hunk 8]
```diff
@@ -446,12 +446,12 @@ var ngOptionsDirective = ['$compile', '$document', '$parse', function($compile,
       };
 
       var removeEmptyOption = function() {
+        emptyOption.removeAttr('selected');
```
Comment (gkalpak) [inline src/ng/directive/ngOptions.js]: Minor (as it isn't expected to happen often in an app's lifecycle), but this could be moved to an `else` branch (since we only care about the attribute if we aren't removing the element).

File: test/ng/directive/ngOptionsSpec.js [hunk 9]
```diff
@@ -120,6 +120,55 @@ describe('ngOptions', function() {
...
+            var errors = [];
+            if (actual.selected === null || typeof actual.selected === 'undefined' || actual.selected === false) {
+              errors.push('Expected option property "selected" to be truthy');
+            }
+
+            if (msie !== 9 && actual.hasAttribute('selected') === false) {
+              errors.push('Expected option to have attribute "selected"');
+            }
+
+            var message = function() {
```
Comment (gkalpak) [inline test/ng/directive/ngOptionsSpec.js]: This function doesn't seem necessary. (You can just `message: errors.join('\n')`.)

**Comment by gkalpak** [comment]
Do the tests fail now without the `emptyOption.removeAttr('selected')` line?

**Comment by Narretz** [comment]
Yep, the tests fail now, when I remove the remoeAttr code. Thanks for the review!

**Comment by dustin-page** [comment]
Thanks for working on this. However, I noticed it is still an issue in AngularJS 1.5.11. When I use ngOptions with track by the selected attributes for the unselected options are not removed.

ng-options-multiple-selected-attributes

Here is a link to Plnkr example:
http://plnkr.co/edit/tpl:fdTouZ7cjNtUDbeZTYnY?p=preview

**Comment by gkalpak** [comment]
It happens on latest master too. @dustin-page can you please open a new issue. This has probably something to do with how you use `track by`.

**Comment by dromerop** [comment]
using 1.6.8 and got same error when using track by: changing the selected element adds the selected attribute to the current item but doesn't remove it from the old one.
